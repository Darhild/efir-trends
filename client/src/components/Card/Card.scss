@import './../../variables';

.Card { 
    display: block;
    height: 100%;
    text-align: left;
    color: inherit;
    text-decoration: none;
    background-color: var(--color-layout);
    border-radius: 6px;
    cursor: pointer;

    &_size {

        &_small {
            width: 117px;
            background-color: transparent;

            @media(min-width: $screen-l) {
                width: 162px;
            }

            .Card-Thumb {
                height: 175px;
                border-radius: 4px;
                transition: transform 150ms;

                @media(min-width: $screen-l) {
                    height: 243px;
                }
            }
        
            .Card-Thumb:hover {
                transform: translateY(-6px);
            }

            .Card-Content {
                padding-right: 0;
                padding-left: 0;
            }
        }

        &_medium {
            width: calc(((100vw - 386px)/7.5)*1.5 + 5px);
            min-width: 180.5px;
            max-width: 248px;

            .Card-Thumb {
                height: 104px;

                @media(min-width: $screen-l) {
                    height: 140px;
                }
            }
        }

        &_big {
            width: 100%;

            .Card-Thumb {
                height: 184px;

                @media(min-width: $screen-s) {
                    height: 104px;
                }

                @media(min-width: $screen-m) {
                    height: 140px;
                }

                @media(min-width: $screen-l) {
                    height: 210px;
                }
            }

            .Card-Score {
                display: none;
            }
        }

        &_full {
            width: 100%;

            .Card-Thumb {
                height: 184px;

                @media(min-width: $screen-s) {
                    height: 104px;
                }
            }
        }
    }

    &-Thumb {
        position: relative;
        height: 104px;
        background: center no-repeat var(--color-placeholder);
        background-size: cover;
        border-radius: 6px 6px 0 0;
    }  
    
    &-Poster {
        position: absolute;
        right: 10px;
        bottom: -10px;
        width: 100px;
        border-radius: 4px;
    }

    &-Details {
        position: absolute;
        bottom: 8px;
        left: 8px;
        display: flex;
    }

    &-Content {
        padding: 8px 10px 12px;
        overflow: hidden;
        text-align: left;

        &_view_grid{
            display: flex;
            justify-content: space-between;
        }
    }

    &-Title {
        max-height: 35px;
        margin-bottom: 3px;
        overflow: hidden;
        font-size: 13px;
        color: var(--text-primary);
        text-overflow: ellipsis;
        -webkit-line-clamp: 2;
        -webkit-box-orient: vertical;
        word-wrap: break-word;
        word-break: break-word;
        hyphens: auto;
    }

    &-Subtitle {
        overflow: hidden;
        font-size: 12px;
        line-height: 16px;
        color: var(--text-subtitle);
        white-space: nowrap;
        text-overflow: ellipsis;
        -webkit-line-clamp: 1;
    }

    &-Left {
        max-width: calc(100% - 76px);
    }

    &-Right {
        display: flex;
        align-items: flex-end;
        padding-left: 10px;
    }
}

