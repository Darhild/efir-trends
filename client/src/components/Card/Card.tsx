import React, { Component } from 'react';
import classnames from 'classnames';
import './Card.scss';

export interface Props {
    title?: React.ReactNode;
    subtitle?: React.ReactNode;
    details?: React.ReactNode;
    rightContent?: React.ReactNode;
    className?: string;
    img?: string;
    imgView?: string;
    size?: string;
    content_id: string;
    poster?: string;
    bgColor?: string;
}

const changeImageSize = (image: string | undefined) => image && image.replace(/\/orig/, '/400x300');
export default class Card extends Component<Props> {
    public renderContent() {
        const {
            title,
            subtitle,
            rightContent,
        } = this.props;

        const titles = (
            <>
                {title && <div className="Card-Title" >{title}</div>}
                {subtitle && <div className="Card-Subtitle">{subtitle}</div>}
            </>
        );

        if (rightContent) {
            return (
                <div className="Card-Content Card-Content_view_grid">
                    <div className="Card-Left">{titles}</div>
                    <div className="Card-Right">{rightContent}</div>
                </div>
            );
        }

        return <div className="Card-Content">{titles}</div>;
    }

    public render() {
        const {
            className,
            size = 'medium',
            content_id,
            img,
            imgView,
            details,
            poster,
            bgColor,
        } = this.props;

        const cardCn = classnames(
            'Card',
            `Card_size_${size}`,
            bgColor && `Card_bgColor_${bgColor}`,
            className,
        );

        const thumbCn = classnames(
            'Card-Thumb',
            imgView && `Card-Thumb_view_${imgView}`,
        );

        return (
            <a
                className={cardCn}
                href={`https://yandex.ru/efir?from=efir&stream_id=${content_id}`}
                target="_blank"
                rel="noopener noreferrer"
            >
                <div
                    className={thumbCn}
                    style={{ backgroundImage: img ? `url(${changeImageSize(img)})` : '' }}
                >
                    {
                        poster && <img src={changeImageSize(poster)} className="Card-Poster" alt="" />
                    }
                    {
                        details && <div className="Card-Details">{details}</div>
                    }
                </div>
                {this.renderContent()}
            </a>
        );
    }
}
