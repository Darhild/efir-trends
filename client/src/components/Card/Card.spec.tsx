import React from 'react';
import Card from './Card';
import renderer from 'react-test-renderer';

test('Card without props', () => {
    const component = renderer.create(<Card id="0" />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('Card with small size', () => {
    const component = renderer.create(<Card id="0" size="small"/>);

    expect(component.toJSON()).toMatchSnapshot();
});

test('Card with large size', () => {
    const component = renderer.create(<Card id="0" size="large"/>);

    expect(component.toJSON()).toMatchSnapshot();
});
