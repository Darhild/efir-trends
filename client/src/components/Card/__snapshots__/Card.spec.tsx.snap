// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card with large size 1`] = `
<a
  className="Card Card_size_large"
  href="https://yandex.ru/efir?from=efir&stream_id=0"
  rel="noopener noreferrer"
  target="_blank"
>
  <div
    className="Card-Thumb"
    style={
      Object {
        "backgroundImage": "",
      }
    }
  />
  <div
    className="Card-Content"
  />
</a>
`;

exports[`Card with small size 1`] = `
<a
  className="Card Card_size_small"
  href="https://yandex.ru/efir?from=efir&stream_id=0"
  rel="noopener noreferrer"
  target="_blank"
>
  <div
    className="Card-Thumb"
    style={
      Object {
        "backgroundImage": "",
      }
    }
  />
  <div
    className="Card-Content"
  />
</a>
`;

exports[`Card without props 1`] = `
<a
  className="Card Card_size_medium"
  href="https://yandex.ru/efir?from=efir&stream_id=0"
  rel="noopener noreferrer"
  target="_blank"
>
  <div
    className="Card-Thumb"
    style={
      Object {
        "backgroundImage": "",
      }
    }
  />
  <div
    className="Card-Content"
  />
</a>
`;
